### PR TITLE
Swift3

### PR DESCRIPTION
**Pull request by jberkel**
Supersedes #501, #475

### CODE REVIEW

**Comment by jberkel** [comment]
@stephencelis it might make sense to skip #475, unless you want to keep a separate branch for 2.3 compatibility?

**Comment by iLandes** [comment]
It's look great. 

How can I integrate it on a Pod file ?

**Comment by jberkel** [comment]
@iLandes 

try
```ruby
pod 'SQLite.swift', :git => 'https://github.com/stephencelis/SQLite.swift.git',
       :branch => 'swift3-mariotaku'
```

**Comment by cmvolare** [comment]
can you I get a carthage line to integrate this branch into my project? thanks in advance.

**Comment by stephencelis** [comment]
Many thanks to @mariotaku for picking up my branch and getting it across the finish line! And many more thanks to @jberkel for doing an incredible job coordinating this migration. It's a big one!

**Comment by stephencelis** [comment]
> it might make sense to skip #475, unless you want to keep a separate branch for 2.3 compatibility?
>
> Alternatively we could merge #475 first and then have a swift3 branch in parallel, but given the amount of changes this will become a maintenance nightmare.

It'd be nice to tag 2.3 compatibility, though I don't think maintaining both at once is feasible.

File: SQLite/Core/Blob.swift [hunk 1]
```diff
@@ -30,16 +30,16 @@ public struct Blob {
         self.bytes = bytes
     }
 
-    public init(bytes: UnsafePointer<Void>, length: Int) {
-        self.init(bytes: [UInt8](UnsafeBufferPointer(
-            start: UnsafePointer(bytes), count: length
-        )))
+    public init(bytes: UnsafeRawPointer, length: Int) {
+        // TODO correct count
```
Comment (jberkel) [inline SQLite/Core/Blob.swift]: @mariotaku what is needed here?

Comment (mariotaku) [inline SQLite/Core/Blob.swift]: It works in tests and actual app, though I'm not sure whether my changes correct, maybe we could take a look at swift 2.3 branch.

File: CocoaPods/iphoneos-10.0/module.modulemap [hunk 2]
```diff
@@ -0,0 +1,4 @@
+module CSQLite [system] {
+    header "/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS10.0.sdk/usr/include/sqlite3.h"
```
Comment (sam-w) [inline CocoaPods/iphoneos-10.0/module.modulemap]: This hard-coding of the Xcode path causes build errors (Carthage, CocoaPods or just building in Xcode) for all users who don't have their copy of Xcode 8.x at that path.

File: CocoaPods/iphonesimulator-10.0/module.modulemap [hunk 3]
```diff
@@ -0,0 +1,4 @@
+module CSQLite [system] {
+    header "/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator10.0.sdk/usr/include/sqlite3.h"
```
Comment (sam-w) [inline CocoaPods/iphonesimulator-10.0/module.modulemap]: As above, this causes build errors for all users who don't have their copy of Xcode 8.x at that path.

File: CocoaPods/macosx-10.11/module.modulemap [hunk 4]
```diff
@@ -0,0 +1,4 @@
+module CSQLite [system] {
+    header "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/usr/include/sqlite3.h"
```
Comment (sam-w) [inline CocoaPods/macosx-10.11/module.modulemap]: As above, this causes build errors for all users who don't have their copy of Xcode 8.x at that path.

File: CocoaPods/macosx-10.12/module.modulemap [hunk 5]
```diff
@@ -0,0 +1,4 @@
+module CSQLite [system] {
+    header "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk/usr/include/sqlite3.h"
```
Comment (sam-w) [inline CocoaPods/macosx-10.12/module.modulemap]: As above, this causes build errors for all users who don't have their copy of Xcode 8.x at that path.

File: CocoaPods/macosx/module.modulemap [hunk 6]
```diff
@@ -1,4 +1,4 @@
 module CSQLite [system] {
-    header "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/usr/include/sqlite3.h"
+    header "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sqlite3.h"
```
Comment (sam-w) [inline CocoaPods/macosx/module.modulemap]: As above, this causes build errors for all users who don't have their copy of Xcode 8.x at that path.

File: CocoaPods/iphoneos-10.0/module.modulemap (hunk 2 above)
Comment (jberkel) [inline CocoaPods/iphoneos-10.0/module.modulemap]: What do you suggest to do instead?

Comment (stephencelis) [inline CocoaPods/iphoneos-10.0/module.modulemap]: This hard-coding is required for CocoaPods. If you include SQLite.swift as a sub-project or via Carthage it uses an SDK-relative path instead.

I've raised this issue with CocoaPods in the past: https://github.com/CocoaPods/CocoaPods/issues/3942

We're always open to alternative solutions for CocoaPods users, though, and would happily take contributions that solve the problem more elegantly for them!

File: SQLite/Core/Blob.swift (hunk 1 above)
Comment (jberkel) [inline SQLite/Core/Blob.swift]: looks good to me, I added a bunch of tests to very the behaviour

**Comment by jberkel** [comment]
PR  is ready for another round of review.

**Comment by acrookston** [comment]
I was just upgrading my app and something isn't working for watchOS. Wanted to leave a note about it. Will be digging in deeper to understand why.

File: CocoaPods/iphoneos-10.0/module.modulemap (hunk 2 above)
Comment (jberkel) [inline CocoaPods/iphoneos-10.0/module.modulemap]: @stephencelis this also affects Carthage builds – without the modulemap, the `CSQLite` imports cannot be resolved. Even worse, they're still required *after* the framework has been built (see #492).

**Comment by jberkel** [comment]
Since I did not hear any more bug reports/complaints I just went ahead and merged the PR. The modulemap problems for non-CocoaPods users should hopefully be gone now.

**Pull request merged**